### ISSUE TITLE
SDL_main not found on Android

### ISSUE THREAD

**Issue by danginsburg** [libsdl-org/SDL#11162]
I sync'd up recently to commit e31c68427a3aa29db6a8ab8ceb1cdd52d3d7edf5 and after rebuilding for Android hit a failure at startup of the apk that SDL_main was not found in libtestsprite.so.  The following patch fixed it for me:

```
diff --git a/include/SDL3/SDL_main.h b/include/SDL3/SDL_main.h
index 6160e5e36..de9c84739 100644
--- a/include/SDL3/SDL_main.h
+++ b/include/SDL3/SDL_main.h
@@ -567,8 +567,7 @@ extern SDL_DECLSPEC void SDLCALL SDL_GDKSuspendComplete(void);
      *  do *not* use SDL_main_impl.h (because SDL_main() is called from external Java code).
      *  If other platforms like that turn up, add them next to "defined(SDL_PLATFORM_ANDROID)"
      */
-    #if (defined(SDL_MAIN_USE_CALLBACKS) || defined(SDL_MAIN_NEEDED) || defined(SDL_MAIN_AVAILABLE)) && \
-        !defined(SDL_PLATFORM_ANDROID)
+       #if (defined(SDL_MAIN_USE_CALLBACKS) || defined(SDL_MAIN_NEEDED) || defined(SDL_MAIN_AVAILABLE))
 
         /* platforms which main (-equivalent) can be implemented in plain C */
         #include <SDL3/SDL_main_impl.h>
```

**Comment by slouken** [comment]
@DanielGibson, can you take a look?

**Comment by DanielGibson** [comment]
I will.

**Comment by DanielGibson** [comment]
This really shouldn't happen.
Do you `#define SDL_MAIN_HANDLED` (or use `-DSDL_MAIN_HANDLED`)? You shouldn't.
Otherwise I don't see how this could happen.

Here SDL_main.h defines `SDL_MAIN_NEEDED` for Android: https://github.com/libsdl-org/SDL/blob/e31c68427a3aa29db6a8ab8ceb1cdd52d3d7edf5/include/SDL3/SDL_main.h#L81-L90

And here is the define that renames main to SDL_main:
https://github.com/libsdl-org/SDL/blob/e31c68427a3aa29db6a8ab8ceb1cdd52d3d7edf5/include/SDL3/SDL_main.h#L166-L168
which should be active here, as SDL_MAIN_NEEDED is defined.

Last but not least, there's the declaration for the SDL_main() function that should make sure it gets exported etc:
https://github.com/libsdl-org/SDL/blob/e31c68427a3aa29db6a8ab8ceb1cdd52d3d7edf5/include/SDL3/SDL_main.h#L425

**Comment by DanielGibson** [comment]
Also note how the old code in that place didn't include SDL_main.h for Android either:

https://github.com/libsdl-org/SDL/blob/cba77834f240198d9f43f438df3e35f4887e71e0/include/SDL3/SDL_main.h#L564-L572

So I really can't explain this problem, but I also don't have any Android development environment so I can't test it myself.
As far as I know SDL has Android CI builds though, right? Shouldn't they run into the same problem?

**Comment by DanielGibson** [comment]
Ahh shit, now I know, this will most probably be about `SDL_MAIN_USE_CALLBACKS`

Does this also work for you:

```
#if !defined(SDL_MAIN_HANDLED) && !defined(SDL_MAIN_NOIMPL)
    /* include header-only SDL_main implementations ... */
    #if defined(SDL_MAIN_USE_CALLBACKS) || \
       ( (defined(SDL_MAIN_NEEDED) || defined(SDL_MAIN_AVAILABLE)) && \
         !defined(SDL_PLATFORM_ANDROID) )
      // ...
```

SDL_main_impl.h must be always included if `SDL_MAIN_USE_CALLBACKS` is set, no matter what platform is used.
But if `SDL_MAIN_USE_CALLBACKS` is **not** set, it must not be included on Android.

**Comment by madebr** [comment]
> So I really can't explain this problem, but I also don't have any Android development environment so I can't test it myself. As far as I know SDL has Android CI builds though, right? Shouldn't they run into the same problem?

Yes, you can download prebuilt apks at the bottom of the actions page.
https://github.com/libsdl-org/SDL/actions/runs/11283609753
Download the "SDL-android-arm64-apks" artifact

This is a runtime error: the programs build fine, but at runtime `SDL_main` cannot be found because `main` was not renamed to `SDL_main` or the callbacks did not generate a `SDL_main`/`main`.

**Comment by DanielGibson** [comment]
You're right, the compiler can't catch this of course, didn't think of that!

**Comment by danginsburg** [comment]
> @danginsburg: Does this work for you:

Yes, that works.

**Comment by DanielGibson** [comment]
Fix is in #11164
